### ISSUE TITLE
tape.gradient returns None when assign is used in a model for computation of a function

### ISSUE THREAD

**Issue by naji-s** [tensorflow/tensorflow#54532]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Springdale
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.8.9
- Python version: 3.7.10+

**Describe the current behavior**
When a function is calculated based on a weight of a model the gradient can't trace the connection and returns None. More precisely I am trying to define a log_posterior from a model that has multiple weights (a shallow neural network, if you will). But Hamiltonian Monte Carlo does not work because my defined log_posterior does not have a gradient.

**Describe the expected behavior**
TF should return the gradient of the log_posterior function defined.

**Standalone code to reproduce the issue**
 In the example below tape.gradient should return the value of w as the gradient. Here DummyModel's call supposed to replicate a log_postrior function, where I reassign its parameter "w" with "x". 

```
>>> import tensorflow as tf
>>> class DummyModel(tf.keras.Model):
    def __init__(self):
        super().__init__()
        pass
    def build(self, input_shape):
        self.w = self.add_weight('w', shape=(), dtype=tf.float64, initializer=lambda x, dtype:tf.constant(10., dtype=tf.float64))
    def call(self, X):
        return tf.cast(X, tf.float64) * self.w
... 
>>> a = DummyModel()
>>> print("the output of DummyModel for X=1:", a(1.))
the output of DummyModel for X=1: tf.Tensor(10.0, shape=(), dtype=float64)
>>> x = tf.Variable(3., dtype=tf.float64)
>>> with tf.GradientTape() as tape:
    tape.watch([x, a.w])
    a.w.assign(x)
    out = a(15.)
... 
<tf.Variable 'UnreadVariable' shape=() dtype=float64, numpy=3.0>
>>> print(tape.gradient(out, x))
None

```
